reconcile/dyn_traffic_director.py

```diff
@@ -0,0 +1,303 @@
+import logging
+from typing import Dict, List, Union
+import warnings
+
+from reconcile import queries
+from reconcile.utils.config import ConfigNotFound, get_config
+from reconcile.utils.secret_reader import SecretReader
+
+# Dirty hack to silence annoying SyntaxWarnings present as of dyn==1.8.1
+# which will pollute our CLI output
+# PR for fix upstream: https://github.com/dyninc/dyn-python/pull/140
+# This is unlikely to be ever fixed as this repo has not been updated in 4 yrs
+with warnings.catch_warnings():
+    warnings.simplefilter('ignore')
+    from dyn.tm import session as dyn_session  # type: ignore
+    from dyn.tm import services as dyn_services  # type: ignore
+    from dyn.tm import zones as dyn_zones  # type: ignore
+    from dyn.tm.services.dsf import DSFCNAMERecord  # type: ignore
+    from dyn.tm.services.dsf import DSFFailoverChain  # type: ignore
+    from dyn.tm.services.dsf import DSFRecordSet  # type: ignore
+    from dyn.tm.services.dsf import DSFResponsePool  # type: ignore
+    from dyn.tm.services.dsf import DSFRuleset  # type: ignore
+    from dyn.tm.services.dsf import TrafficDirector  # type: ignore
+
+QONTRACT_INTEGRATION = 'dyn-traffic-director'
+
+DEFAULT_TD_TTL = 30
+DEFAULT_WEIGHT = 100
+
+
+class InvalidRecord(Exception):
+    pass
+
+
+class UnsupportedRecordType(Exception):
+    pass
+
+
+def fetch_current_state() -> dict:
+    state: dict = {
+        'tds': {},
+    }
+
+    for td in dyn_services.get_all_dsf_services():
+        td_name: str = td.label
+        td_nodes: List[str] = [node['fqdn'] for node in td.nodes
+                               if 'fqdn' in node]
+        td_ttl: int = td.ttl
+
+        records: List[Dict[str, Union[str, int]]] = []
+        # Need to follow the hierarchy, even though the one we create/manage
+        # is flat (only one of each)
+        # TD <- Ruleset -> ResponsePool <- FailoverChain <- RecordSet <- Record
+        for ruleset in td.rulesets:
+            for pool in ruleset.response_pools:
+                for chain in pool.rs_chains:
+                    for recordset in chain.record_sets:
+                        for record in recordset.records:
+                            rdata = record.rdata()
+                            if 'cname_rdata' not in rdata:
+                                raise UnsupportedRecordType(
+                                    f'record type {type(record)} is not '
+                                    f'supported: {record}'
+                                )
+
+                            cname_rdata = record.rdata()['cname_rdata']
+                            if 'cname' not in cname_rdata:
+                                raise UnsupportedRecordType(
+                                    f'record missing a cname field: {record}'
+                                )
```

Comment:
The integration supports only CNAME records. We're not going to manage existing TD configs

The idea is if it ever comes across a record type it can't understand it will error out and we can intervene